src/hooks/useUpdateSearchParams.ts

```diff
@@ -0,0 +1,59 @@
+import { useSearchParams } from 'react-router-dom';
+
+type UpdateSearchParamsOptions = {
+  method?: 'push' | 'replace';
+};
+
+type UpdateSearchParams = (
+  key: string,
+  value: string | null,
+  options?: UpdateSearchParamsOptions,
+) => void;
+
+/**
+ * This hook provides an `updateSearchParams` function that can update the
+ * search paramters _without_ creating a rerender on the router level.
+ *
+ * NOTE: It's important to handle the search param you're adding in the react
+ * router loader to make sure that loading a URL from browser or refreshing the
+ * page in browser handles it properly.
+ */
+const useUpdateSearchParams = (): UpdateSearchParams => {
+  // useSearchParams hook provides the current search parameters
+  // and a method to update them, which would trigger a re-render.
+  const [searchParams] = useSearchParams();
```

Comment:
Frankly I'd bail out of everything related to react router in here. So I'd get the searchParams from `location.search` directly, and I'd also expose them like the original react-router hook.

This would have the nice following advantages:
* it could be used anywhere and not just at the top level of a react component
* it wouldn't subscribe to the react router context -- I think that when using `useSearchParams` we'd get a rerender if _something somewhere_ changes the location.
* A developer could still use both hooks (this one and the one from react router) together to be able to get this behavior is they want to.
* by returning the `searchParams` from this hook, I believe we don't need to change the loader because the render function could pick it up when rendering. Maybe this is true only for `replaceState` though? I'm not sure.

Then I'd call the hook `useRawSearchParams` (maybe)?